### PR TITLE
feat(14248): add returns tag to JSDoc template

### LINKED ISSUE #14248: JSDoc generation - generate @returns when function returns a value
_From @egamma on February 22, 2017 11:14_

In the following generated JS Doc comment I would have expected that there is also a `@returns` comment.
 
```js
/**
 * 
 * @param {any} a
 * @param {any} b
 */
function foo(a, b) {
  return 12;
}
```

_Copied from original issue: Microsoft/vscode#21109_

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #14248

### CODE REVIEW

File: src/services/jsDoc.ts [hunk 1]
```diff
@@ -269,34 +269,35 @@ namespace ts.JsDoc {
...
-        return { newText: result, caretOffset: preamble.length };
+        const openComment = "/**";
+        const closeComment = " */";
+        if (tags) {
+            const preamble = openComment + newLine + indentationStr + " * ";
+            const endLine = tokenStart === position ? newLine + indentationStr : "";
+            const result = preamble + newLine + tags + indentationStr + closeComment + endLine;
+            return { newText: result, caretOffset: preamble.length };
+        }
+        return { newText: openComment + closeComment, caretOffset: 3 };
```
Comment (jessetrinity) [inline src/services/jsDoc.ts:300]: The `caretOffset` was previously calculated using `newLine` which is system dependent. Seems like changing it would cause problems, no?

File: src/services/jsDoc.ts [hunk 2]
```diff
@@ -315,9 +316,14 @@ namespace ts.JsDoc {
         }).join("");
     }
 
+    function returnsDocComment(isJavaScriptFile: boolean, indentationStr: string, newLine: string) {
+        return `${indentationStr} * @returns ${isJavaScriptFile ? "{any}" : ""}${newLine}`;
```
Comment (jessetrinity) [inline src/services/jsDoc.ts:320]: We should have a test for `@returns {any}` in JS files if we add it.

**Review by jessetrinity — CHANGES_REQUESTED** [review]
I agree with adding `@returns` without the type for `.ts` files. @sandersn based on the usage analysis you did it seemed like few JSDoc comments in .js make use of `@returns` (though those who did almost always included the type). Is there an argument against the `.js` case? Thanks!

File: src/services/jsDoc.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/services/jsDoc.ts:300]: The following code that was used before

```ts
const singleLineResult = "/** */";	
return { newText: singleLineResult, caretOffset: 3 };	
```

File: src/services/jsDoc.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/services/jsDoc.ts:320]: @jessetrinity Done.

File: src/services/jsDoc.ts (hunk 1 above)
Comment (jessetrinity) [reply src/services/jsDoc.ts:300]: ah whoops.

**Review by jessetrinity — APPROVED** [review]
Looks good. I think this is the way to go but I'd like to hear opinions from @sandersn or @DanielRosenwasser as we had a lot of offline discussion about the inclusion/exclusion of the type.

**Comment by jessetrinity** [comment]
@a-tarasyuk after talking at length with @DanielRosenwasser It seems I need to backpedal a bit. It seems adding `@returns {any}` in any case would be undesirable.

1. We don't want `{any}` under `// @ts-check` 
2. `@returns {any}` annotations would clobber type inference otherwise.
3. We don't want this operation to depend on doing semantic work as the syntax server should only handle very fast operations. Waiting on other semantic operations to complete to add the JSDoc scaffolding could have a negative impact on responsiveness.

TL;DR we should only add `@returns` instead of `@returns {any}`. Sorry for the runaround.

**Review by jessetrinity — CHANGES_REQUESTED** [review]
@a-tarasyuk after talking at length with @DanielRosenwasser It seems I need to backpedal a bit. It seems adding `@returns {any}` in any case would be undesirable.

1. We don't want `{any}` under `// @ts-check` 
2. `@returns {any}` annotations would clobber type inference otherwise.
3. We don't want this operation to depend on doing semantic work as the syntax server should only handle very fast operations. Waiting on other semantic operations to complete to add the JSDoc scaffolding could have a negative impact on responsiveness.

TL;DR we should only add `@returns` instead of `@returns {any}`. Sorry for the runaround.

**Review by jessetrinity — APPROVED** [review]
Thanks!

**Comment by bes-aslan** [comment]
> 
> 
> I agree with adding `@returns` without the type for `.ts` files. @sandersn based on the usage analysis you did it seemed like few JSDoc comments in .js make use of `@returns` (though those who did almost always included the type). Is there an argument against the `.js` case? Thanks!

@jessetrinity, this is a really useful feature - many thanks to all involved. Will this feature also cover auto-generating @returns JSDoc for .js files as well (as .ts files)? Or should I be contacting someone else about that?
